Pathfind fix for NPC/MOB and poor old Novalmauge

Omnione opened the pull request:
• Addition of npcBasicPath to pathfind.lua, this pathfind function requires paths to be set in a linear fashion like this example
local path =
{
    [1] = {1.000, 2.000, 3.000},
    [2] = {2.000, 2.000, 3.000},
    [3] = {3.000, 2.000, 3.000},
    [4] = {2.000, 2.000, 3.000}
}
This way the path returns to the origin from the last path point to the first, i.e a circle.
the npcBasicPath will keep track of what path point the npc is on by the newly added m_pathPoint in npcentity.

• As stated above i have added a m_pathPoint to npcentity and given the the entity a getter and setter to help keep track of the npc's path.

• Adjusted Novalmauge's script to reflect the changes.

He's walk animation still seems a bit buggy but its about as good as i could get it.
But atleast it shouldn't crash the server anymore.
 
This should fix the following bug
Novalmauge causes server crash due to pathing. #493

<!-- place 'x' mark between square [] brackets to affirm: -->
**_I affirm:_**
- [X] that I agree to Project Topaz's [Limited Contributor License Agreement](http://project-topaz.com/blob/release/CONTRIBUTOR_AGREEMENT.md), as written on this date
- [X] that I've _tested my code_ since the last commit in the PR, and will test after any later commits

Omnione committed:
Pathfind fix for Novalmauge

• Addition of npcBasicPath to pathfind.lua, this pathfind function requires paths to be set in a linear fashion like this example
local path =
{
    [1] = {1.000, 2.000, 3.000},
    [2] = {2.000, 2.000, 3.000},
    [3] = {3.000, 2.000, 3.000},
    [4] = {2.000, 2.000, 3.000}
}
This way the path returns to the origin from the last path point to the first, i.e a circle.
the npcBasicPath will keep track of what path point the npc is on by the newly added m_pathPoint in npcentity.

• As stated above i have added a m_pathPoint to npcentity and given the the entity a getter and setter to help keep track of the npc's path.

• Adjusted Novalmauge's script to reflect the changes.

Omnione committed:
Pathfind fix for Novalmauge

• Addition of npcBasicPath to pathfind.lua, this pathfind function requires paths to be set in a linear fashion like this example
local path =
{
    [1] = {1.000, 2.000, 3.000},
    [2] = {2.000, 2.000, 3.000},
    [3] = {3.000, 2.000, 3.000},
    [4] = {2.000, 2.000, 3.000}
}
This way the path returns to the origin from the last path point to the first, i.e a circle.
the npcBasicPath will keep track of what path point the npc is on by the newly added m_pathPoint in npcentity.

• As stated above i have added a m_pathPoint to npcentity and given the the entity a getter and setter to help keep track of the npc's path.

• Adjusted Novalmauge's script to reflect the changes.

Changed setPathPoint to use uint16 rather than uint8

Stupid tabs

Pathing adjustments

• Added the ability to pass rotations
• Moved m_pathPoint to baseEntity

Omnione committed:
Changed setPathPoint to use uint16 rather than uint8

Omnione committed:
Pathing adjustments

• Added the ability to pass rotations
• Moved m_pathPoint to baseEntity

ibm2431 commented on scripts/zones/Bostaunieux_Oubliette/npcs/Novalmauge.lua:
```diff
@@ -15,39 +15,38 @@ require("scripts/globals/quests")
-    51.917370, -23.924366, 19.970068,
-    74.570229, -24.024828, 20.103880,
-    44.533886, -23.947662, 19.926519
+    [1] = {20.000, -24.000, 20.000, 0},
+    [2] = {50.000, -24.000, 20.000, 0},
+    [3] = {75.000, -24.000, 20.000, 0},
+    [2] = {50.000, -24.000, 20.000, 0}
```
123..2?

In either case, the table keys aren't needed; Lua will internally give un-keyed elements an increasing integer representation 😉 

(Unless you've changed the key structure with hard-indexed elements, but that's a different can of worms.)

Omnione replied:
yeah true, i guess i got so used to doing table like this for readability, that i just did it this way.

Omnione replied:
all tables that use the new system should now be a simple point i.e. {10, 10, 10} with additional extra params able to be passed if required.

ibm2431 commented on src/map/lua/lua_baseentity.cpp:
```diff
@@ -1741,6 +1741,46 @@ inline int32 CLuaBaseEntity::clearTargID(lua_State* L)
+inline int32 CLuaBaseEntity::setPathPoint(lua_State* L)
+{
+    TPZ_DEBUG_BREAK_IF(m_PBaseEntity == nullptr);
+    TPZ_DEBUG_BREAK_IF(m_PBaseEntity->objtype != TYPE_NPC);
```
If change the break to be only on players, we might be able to use this pathing method for mobs too~!

Omnione replied:
yeah sounds good to me

Omnione replied:
This has been done now

ibm2431 commented on scripts/zones/Bostaunieux_Oubliette/npcs/Novalmauge.lua:
```diff
@@ -15,39 +15,38 @@ require("scripts/globals/quests")
 function onSpawn(npc)
     npc:initNpcAi()
-    npc:setPos(tpz.path.first(path))
+    npc:setPos(path[1][1], path[1][2], path[1][3], 0)
+    npc:speed(12)
+    npc:setPathPoint(1)
```
Thoughts of baking this into `initNpcAi`? Can make it take a table, grab the first point while in core, set the pos, and also potentially take a walking speed as well to set as a local var.

Omnione replied:
i did change the name of initNpcAI to initNpcPathing and you can now call it as npc:initNpcPathing() or pass an x,y,z to set the start position so npc:initNpcPathing(x,y,z) and also it will set the path point to 1 as default, but you can pass a 4th param to set the point also so npc:initNpcPathing(x, y, z, 5) will set the position and set the starting path point to 5.

ibm2431 commented on scripts/zones/Bostaunieux_Oubliette/npcs/Novalmauge.lua:
```diff
@@ -15,39 +15,38 @@ require("scripts/globals/quests")
     if player:getCharVar("troubleAtTheSluiceVar") == 2 and npcUtil.tradeHas(trade, 959) then -- Dahlia
         player:startEvent(17)
-        npc:wait()
+        npc:speed(0)
```
If all paroling NPCs use this new method, will `npc:wait()` still have a use? Should we consider changing the internals of `:wait()` to essentially set speed to 0? (Would pathing need to be stopped?)

If we need to keep `npc:wait()`, should there be a wrapper like `tpz.path.pause()`?

TeoTwawki replied:
opinion: wait() does a lot of things besides stopping movement, it stops all the stuff and things. because of that its tempting to use it for a lot of situations that result in crashes. wait should be removed from its use in pathing scripts, but may have future use for..something.. 

could we possibly cause the event triggering core side to set the entities speed to 0 and then the event finish to restore it to its original value? :thinking: `m_event.Target` appears to be the npc in the core.

ibm2431 replied:
> m_event.Target appears to be the npc in the core

I was not aware of this - we can make event starts accurate! (On retail, the packet is the initiating NPC, but we've always invoked it directly as the "player" being the initiator)

Omnione replied:
I'm not really sure where to go with this but as we now have pathStop() and pathResume(), we will no longer get crashes when interacting with an npc that uses pathing, i think i found and converted all npc's that use pathing to use the new pathing systems.

Omnione replied:
I can look into making event set the speed to 0 at a later date if still wanted.

ibm2431 commented on scripts/zones/Bostaunieux_Oubliette/npcs/Novalmauge.lua:
```diff
@@ -109,5 +108,6 @@ function onEventFinish(player, csid, option, npc)
     end
 
-    npc:wait(0)
+    npc:speed(12)
```
If the "new" initNpcAi (which has a commented TODO to change the name) ends up setting a local var to the NPC, then could add a `tpz.path.resume()` which sets the speed back and calls `npcBasicPath`

Omnione replied:
sounds good to me, seems a better way of doing it.

Omnione replied:
Might aswell hold this till that gets implimented if thats what you want to do, then i can change it, might have to give me a nudge to change it incase i forget lol.

Omnione replied:
implimented pathStop() and pathResume() to move away from using wait()

ibm2431 reviewed (commented):
General question: I assume that setting speed to 0 actually does cease movement and doesn't cause sliding? I remember that being a question in #319

Omnione committed:
Big update

Added new pathing options for entity's
First up: pathfind.lua, probly quicker to look at the commenting than me explain that,

changed amost npc's that use pathing to use either basicPath or advancedPath, the advancedPath is the one that stands out as this allows the entity to pass id text and npc id's to automatically look at or speak at a a point in the path, and each point can have a delay.

Added random pathing for the bunnies in Khazam, as they just sat there.

I kept everything the same where i could, but i did have to redo some paths, but that let me cut down on path points also.

Omnione committed:
Changes and additions

• Added a new binding rotateToAngle(), rotates an entity to a given angle, NOTE: must not be moving.

• added the ability to add extra params to a path point when using tpz.path.general, example: 
basic point {x,y,z} 
point with a delay {x,y,z {delay = 10}}
point with delay and rotation {x,y,z, {rot = 35, delay = 10}
point with a rotation only {x,y,z {rot = 120}}, in this case a small delay will automatically be set so the rotation can be applied.

• Updated all npc's that now use the new lua path functions and tested them.

zach2good commented on scripts/zones/Davoi/npcs/Quemaricond.lua:
```diff
@@ -4,58 +4,57 @@
-    53.5, -1.8, -19,
-    61, -1.1, -18.6,
-    67.3, -1.5, -18.6,
-    90, -0.5, -19
-};
+    {20.600, -0.343, -23.000, {delay = 5}},
+    {35.431, -0.045, -20.227},
+    {51.795, -1.790, -18.843},
+    {58.037, -0.957, -18.724},
+    {65.919, -1.722, -18.767},
+    {92.143, -0.451, -16.231, {delay = 5}},
+    {65.919, -1.722, -18.767},
+    {58.037, -0.957, -18.724},
+    {51.795, -1.790, -18.843},
+    {35.431, -0.045, -20.227},
+}
```
This section really nicely highlights the benefits of this PR!

zach2good commented on scripts/zones/Davoi/npcs/Quemaricond.lua:
```diff
@@ -4,58 +4,57 @@
+end
 
 function onPath(npc)
+    tpz.path.general(npc, path, tpz.path.flag.WALLHACK, false)
```
Looking at this "in the wild", I wonder if `general` is really the right name for the default, regular, general call?

Now that I say the other options, I guess it's as good as all the others 🤷

zach2good replied:
Maybe refactor to `tpz.pathing.path()`, I don't know how much work that would be?

zach2good commented on scripts/zones/Kazham/npcs/Kukupp.lua:
```diff
@@ -3,75 +3,30 @@
-    41.515419, -11.000000, -177.255875,
-    40.609776, -11.000000, -176.645706
-};
+    {44.000, -11.000, -176.000},
+    {47.000, -11.000, -179.500},
+    {44.000, -11.000, -177.000},
+    {40.500, -11.000, -179.500},
+    {39.500, -11.000, -176.500},
+    {41.300, -11.000, -172.300},
+    {45.000, -11.000, -171.300},
+    {46.000, -11.000, -173.600},
+    {42.700, -11.000, -174.300},
+    {44.000, -11.000, -179.800},
```
I hope you were able to automate this process and didn't have to manage all these tables by hand...

zach2good commented on src/map/lua/lua_baseentity.h:
```diff
@@ -108,10 +108,17 @@ class CLuaBaseEntity
     int32 pathThrough(lua_State* L);         // walk at normal speed through the given points
-    int32 isFollowingPath(lua_State* L);     // checks if the entity is following a path
+    int32 isFollowingPath(lua_State* L);
+	int32 rotateToAngle(lua_State* L);
```
Theres some borked tab indents around here

zach2good reviewed (approved):
So, it all looks good. There are some styling things I'd like to see fixed and in an ideal world, this would have been much easier to go through and review if you had split out the commits to be logic changes, whitespace/semicolon changes, tabling changes etc.

Sick work 👍

zach2good commented:
I think you've taken care of all semicolons and random things in Lua, but if you could take a merge of release just to be sure that would be grand

TeoTwawki commented on src/map/lua/lua_baseentity.cpp:
```diff
@@ -1878,6 +1996,100 @@ inline int32 CLuaBaseEntity::isFollowingPath(lua_State* L)
+        {
+            auto PNpc = static_cast<CNpcEntity*>(m_PBaseEntity);
+            uint32 npcID = PNpc->id;
+            const char* NPCQuery = "SELECT speed FROM npc_list WHERE npcid = %u";
```
Why not store the original speed in a localVar instead of re-querying the database?
You can actually get and set localVars from the C++ side too, not just in lua.

TeoTwawki replied:
For NPCs we could try having the event code itself stop and restart NPC movement. since `m_event.target` is the NPC entity itself. Mobs we'd still need a script command for stop/start.

TeoTwawki commented on src/map/lua/lua_baseentity.cpp:
```diff
@@ -1878,6 +1996,100 @@ inline int32 CLuaBaseEntity::isFollowingPath(lua_State* L)
+    TPZ_DEBUG_BREAK_IF(m_PBaseEntity == nullptr);
+    TPZ_DEBUG_BREAK_IF(m_PBaseEntity->objtype == TYPE_PC);
+
+    CBattleEntity* PBattle = (CBattleEntity*)m_PBaseEntity;
```
`m_PBaseEntity->speed = 0;` actually does work, no need to recast as a battleentity for either mobs or npc's (or players). Which makes this entire function in its current form a duplicate of `speed()`

Lets improve it by having this function also store the entities current speed to a localVar prior to setting it zero. Then resume path can read that value instead of querying the database every time.

```c++
 m_PBaseEntity->SetLocalVar(lua_tostring(L, 1), (uint64_t)lua_tointeger(L, 2));
```

59blargedy replied:
do we want to place all these in an issue or ?

TeoTwawki replied:
not particularly no. I'd rather see it be changed here and its such a small correction to make instead of unnecessarily adding more iops of our inefficient db connection.

zach2good reviewed (changes requested):
I think @Wiggo32 had some things to say about this, putting on hold until they can chime in
